src/blob-store/index.js

```diff
@@ -137,7 +137,13 @@ export class BlobStore extends TypedEmitter {
     const entriesStream = createEntriesStream(this.#driveIndex, { live })
     if (!filter) return entriesStream
     const filterStream = new FilterEntriesStream(filter)
-    return pipeline(entriesStream, filterStream, noop)
+    const result = pipeline(entriesStream, filterStream, noop)
+    // Destroying the pipeline should destroy the *first* stream, not the
+    // whole pipeline.
+    Object.defineProperty(result, 'destroy', {
+      value: entriesStream.destroy.bind(entriesStream),
+    })
+    return result
```

Comment:
I don't quite understand why this is needed? `result` here is actually `filterStream`. My understanding of `pipeline` is that it ensures that destroying any stream in the pipeline also destroys the others. This seems to be overriding the `destroy` method of `filterStream` to destroy `entriesStream`. What is the intent of this?

Reply:
I don't love this solution, but it was the best I could come up with.

First, as a given: we need to be able to destroy the result of this function. Here's that code:

https://github.com/digidem/comapeo-core/blob/7c810c217fc08e25a5ba1c9ce7ab684981b1cde5/src/mapeo-project.js#L438

We need to be able to destroy `entriesStream` specifically. Calling `destroy()` on the result of `pipeline()` causes an error (I recall something like "premature close").

I considered a few other options, none of which I liked:

- Change the requirements and *don't* destroy the stream. I don't think this is a good idea because it'd leave streams around.
- Call `destroy()` on the result of `pipeline()` and handle the rejection. I don't think this actually cleans things up, but it stops the error, I think?
- Return multiple values; maybe something like `{ resultStream, destroyEntriesStream }` or something. I think that muddies the consumer's API, but it would also work. Also, what would it look like if we didn't include a filter?

Hope this answers the question.

Reply:
Thanks, got it. So, a few things (streams are an absolute headache):

1. `ERR_STREAM_PREMATURE_CLOSE` is an expected error from destroying a pipeline before it is finished, this is fine and should be handled.
2. Normally when creating a stream like [you do here](https://github.com/digidem/comapeo-core/blob/7c810c217fc08e25a5ba1c9ce7ab684981b1cde5/src/mapeo-project.js#L438), you would need to attach an `error` listener, to avoid uncaught errors. However, iterating over the stream attaches the error listener for you, and throws with it inside the iterator, so you don't need to attach an error listener here.
3. You do need to wrap the async iterator of `entriesReadStream` with `try {} catch {}` and check `error.code === 'ERR_STREAM_PREMATURE_CLOSE'` - you can ignore that error, but other errors should be logged.
4. You also need to clean up the stream references once the entriesReadStream closes, via `entriesReadStream.on('close', () => if (entriesReadStreams.get(peerId) === entriesReadStream) entriesReadStream.delete(entriesReadStream))`
5. You can remove the checks on `entriesReadStream.destroyed` in the iterator - catching the iterator being thrown (with `ERR_STREAM_PREMATURE_CLOSE`) is the correct way to handle a stream being destroyed.

Reply:
TL;DR:

1. Remove the `Object.defineProperty` code
2. Catch errors in the iterator that reads the stream
3. Remove references to closed streams

Reply:
Edit: add cleanup after the iterator loop rather than on the close event (it's the same thing, just avoids mixing two syntaxes).

Reply:
Thanks for this; I'm bookmarking it for later! Fixed in 13f9a3d882604252994f86bbabb956e88d6b2ded.